RNNLM lattice rescoring with Tensorflow

hainan-xv opened the pull request.

vince62s commented:
Hey Hainan,
This is great. Do you have some timing benchmark both for training and decoding ?
When I did LM rescoring with a TF rnnlm it was very slow and infeasible in cpu, only on gpu.
cheers, great work.

hainan-xv commented:
I did a comparison between TF vanilla-RNNLM lattice rescoring with CUED-RNNLM lattice rescoring (https://github.com/hainan-xv/kaldi/tree/cuedrnnlm, where I re-implemented the underlying matrix operations to use Kaldi matrix instead). The size of the networks are the default values in egs/ami/local/tensorflow/* . For CUED-RNNLM I used O2 optimization and for TF, I used -c opt --copt=-msse4.1 --copt=-msse4.2 to compile the TF libraries, and then use O2. It seems TF is around 2-3 times faster than CUED-RNNLM, when I limit the rescoring n-gram order to be 3.
It seems TF-LSTM adds around 20% time compared to vanilla-rnnlms. 

From my experience, it's crucial that you turn on the compilation optimizations for TF - use "--config=opt " when you can. The choice of me using only "--copt=-msse4.1 --copt=-msse4.2 " is because of the limitation of the CLSP grid. Running rescoring without these options seem much much slower. Add the options in this file https://github.com/hainan-xv/kaldi/blob/tensorflow-rnnlm/tools/install_tensorflow_cc.sh.

danpovey commented:
I assume that's TF without GPU?  And you think you were comparing networks
of about the sane size?


On Tue, Jul 11, 2017 at 5:52 PM, hainan-xv <notifications@github.com> wrote:

> I did a comparison between TF vanilla-RNNLM lattice rescoring with
> CUED-RNNLM lattice rescoring (https://github.com/hainan-xv/
> kaldi/tree/cuedrnnlm, where I re-implemented the underlying matrix
> operations to use Kaldi matrix instead). The size of the networks are the
> default values in egs/ami/local/tensorflow/* . For CUED-RNNLM I used O2
> optimization and for TF, I used -c opt --copt=-msse4.1 --copt=-msse4.2 to
> compile the TF libraries, and then use O2. It seems TF is around 2-3 times
> faster than CUED-RNNLM, when I limit the rescoring n-gram order to be 3.
> It seems TF-LSTM adds around 20% time compared to vanilla-rnnlms.
>
> From my experience, it's crucial that you turn on the compilation
> optimizations for TF - use "--config=opt " when you can. The choice of me
> using only "--copt=-msse4.1 --copt=-msse4.2 " is because of the limitation
> of the CLSP grid. Running rescoring without these options seem much much
> slower. Add the options in this file https://github.com/hainan-xv/
> kaldi/blob/tensorflow-rnnlm/tools/install_tensorflow_cc.sh.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-314582963>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu_ZR60i1YNDLQi0Y5cEqa8HOfHX1ks5sM-6QgaJpZM4OO-zM>
> .
>

danpovey commented:
Also how do the WERs compare?

On Tue, Jul 11, 2017 at 5:55 PM, Daniel Povey <dpovey@gmail.com> wrote:

> I assume that's TF without GPU?  And you think you were comparing networks
> of about the sane size?
>
>
> On Tue, Jul 11, 2017 at 5:52 PM, hainan-xv <notifications@github.com>
> wrote:
>
>> I did a comparison between TF vanilla-RNNLM lattice rescoring with
>> CUED-RNNLM lattice rescoring (https://github.com/hainan-xv/
>> kaldi/tree/cuedrnnlm, where I re-implemented the underlying matrix
>> operations to use Kaldi matrix instead). The size of the networks are the
>> default values in egs/ami/local/tensorflow/* . For CUED-RNNLM I used O2
>> optimization and for TF, I used -c opt --copt=-msse4.1 --copt=-msse4.2 to
>> compile the TF libraries, and then use O2. It seems TF is around 2-3 times
>> faster than CUED-RNNLM, when I limit the rescoring n-gram order to be 3.
>> It seems TF-LSTM adds around 20% time compared to vanilla-rnnlms.
>>
>> From my experience, it's crucial that you turn on the compilation
>> optimizations for TF - use "--config=opt " when you can. The choice of me
>> using only "--copt=-msse4.1 --copt=-msse4.2 " is because of the limitation
>> of the CLSP grid. Running rescoring without these options seem much much
>> slower. Add the options in this file https://github.com/hainan-xv/k
>> aldi/blob/tensorflow-rnnlm/tools/install_tensorflow_cc.sh.
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-314582963>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ADJVu_ZR60i1YNDLQi0Y5cEqa8HOfHX1ks5sM-6QgaJpZM4OO-zM>
>> .
>>
>
>

hainan-xv commented:
Yes it's on CPUs, and limiting to use one CPU. The network sizes are comparable.

danpovey commented:
And the vocab sizes were the same?  Let me know about the WERs too.


On Tue, Jul 11, 2017 at 5:56 PM, hainan-xv <notifications@github.com> wrote:

> Yes it's on CPUs, and limiting to use one CPU. The network sizes are
> comparable.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-314583980>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuzRotOmTzYPpl5ghpFV4YntEyIY2ks5sM--hgaJpZM4OO-zM>
> .
>

hainan-xv commented:
Yes same vocab size.

I didn't tune the vanilla network for PPL and it was mostly for comparing speed.
I'm getting better results with LSTMs, both in terms of PPLs (much better) and WERs (slightly better than vanilla) in rescoring.

hainan-xv commented:
Oh, actually RE that 2-3 times speed up on TF:

I remember that I came up with a more efficient algorithm for
lattice-rescoring than the previous implementation - the old implementation
in CUED-RNNLM, in each FST state (corresponding to history a b c) it
remembers the hidden representation of (a b) and have to recompute the
hidden-representation of (a b c) every time an outgoing arc needs to be
computed.

In the TF implementation I'm instead storing the hidden representation of
(a b c), and this saves some duplicate of computation, and IIRC around 3-4
times speedup. So if taken that into account, TF is slightly slower.

On Tue, Jul 11, 2017 at 2:58 PM, Daniel Povey <notifications@github.com>
wrote:

> And the vocab sizes were the same? Let me know about the WERs too.
>
>
> On Tue, Jul 11, 2017 at 5:56 PM, hainan-xv <notifications@github.com>
> wrote:
>
> > Yes it's on CPUs, and limiting to use one CPU. The network sizes are
> > comparable.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-314583980>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ADJVuzRotOmTzYPpl5ghpFV4YntEyIY2ks5sM--hgaJpZM4OO-zM>
>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-314584248>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFMCDgKavqCYI1oY1mGyE38Ls2UuClMdks5sM-_qgaJpZM4OO-zM>
> .
>



-- 
- Hainan

hainan-xv commented:
The logic becomes slightly more complex for LSTM than vanilla RNN because we will need to store 2 tensors for each state, but it greatly reduces running time.

vince62s commented:
Can you please be specific wrt the vcoab size and network size.
If default is 10k vocab and 2 layers of size 200, yes it should be fast.
But I doubt you will get very good results, and for real world (ie tedlium like corpus)
i had to go up to 2 layers of 1024 with a vocab of 150k

We will run your code for tedlium.

hainan-xv commented:
Yes the default is 10k vocab, one hidden-layer of 200. All the detailed information is in egs/ami/local/tensorflow/*

For larger vocabs, I would recommend running steps/tensorflow/lstm_fast.py . It used a modified objf which automatically normalizes the output, and so there is no need to do softmax over all words.

danpovey commented on tools/install_tensorflow_cc.sh:
```diff
@@ -0,0 +1,52 @@
+
+exit
+
+# the following would utilize the highest optimization but might not work in a
```
Using "exit" to comment code is very confusing.  Just comment the lines and remove the "exit".

danpovey commented on tools/install_tensorflow_cc.sh:
```diff
@@ -0,0 +1,52 @@
+
+set -e
+
+export JAVA_HOME=/export/b02/hxu/TensorFlow/java/jdk1.8.0_121
```
I don't think this export statement belongs here, at least not in the checked-in version.

danpovey commented on tools/install_tensorflow_py.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
```
These kinds of scripts should be in tools/extras/, not in tools/.

danpovey replied:
I don't see either of these install_tensorflow scripts referred to anywhere else in your PR.
They should at least be mentioned in an error message with installation instructions, for when
the example scripts are invoked but TensorFlow was not installed.

danpovey commented on src/tensorflow/tensorflow-rnnlm-lib.cc:
```diff
@@ -0,0 +1,361 @@
+  num_rnn_words = rnn_label_to_word_.size();
+  
+  // we must have a oos symbol in the wordlist
+  if (oos_ == -1) {
```
no need for curly braces here.

danpovey commented on src/tensorflow/tensorflow-rnnlm-lib.cc:
```diff
@@ -0,0 +1,361 @@
+  }
+}
+
+KaldiTfRnnlmWrapper::KaldiTfRnnlmWrapper(
```
The comments in this code are not very helpful, I think it needs to be better-commented or refactored.

danpovey commented on src/tensorflow/tensorflow-rnnlm-lib.h:
```diff
@@ -0,0 +1,156 @@
+  KALDI_DISALLOW_COPY_AND_ASSIGN(KaldiTfRnnlmWrapper);
+};
+
+class TfRnnlmDeterministicFst
```
put the colon directly after TfRnnlmDeterministicFst, like in  other Kaldi code.

danpovey commented on src/tensorflow/tensorflow-rnnlm-lib.h:
```diff
@@ -0,0 +1,156 @@
+// tensorflow-rnnlm-lib.h
```
I don't like '-lib' appearing in the name of these C++ files, it could cause some confusion.  Just remove the -lib.

danpovey commented on src/tensorflow/tensorflow-rnnlm-lib.h:
```diff
@@ -0,0 +1,156 @@
+
+// Copyright         2017 Hainan Xu
+
+// wrapper for tensorflow rnnlm
```
this comment doesn't belong here.

danpovey commented on src/tensorflow/tensorflow-rnnlm-lib.h:
```diff
@@ -0,0 +1,156 @@
+// tensorflow-rnnlm-lib.h
+
+// Copyright         2017 Hainan Xu
```
check with Yishay what the copyright header should say.

danpovey commented on src/tensorflow/tensorflow-rnnlm-lib.h:
```diff
@@ -0,0 +1,156 @@
+                       Tensor *context_out,
+                       Tensor *new_cell);
+
+  // since usually we have a smaller vocab in RNN than the whole vocab,
```
need more explanation about this variable, e.g. are the RNN labels  zero based?  What about FST labels that are OOS?  You need to explain what OOS means, too, when uyou use the term.

danpovey commented on src/tensorflow/tensorflow-rnnlm-lib.h:
```diff
@@ -0,0 +1,156 @@
+
+  KaldiTfRnnlmWrapper *rnnlm_;
+  int32 max_ngram_order_;
+  std::vector<tensorflow::Tensor> state_to_context_;
```
You should probably add an extra layer of indirection here, e.g. pointers to Tensors.  Otherwise when this vector gets resized, it could end up doing a lot of work, depending how the standard library is implemented.

hainan-xv replied:
I think I know why you're suggesting this, but it seems in the case of TF it'll make the code a lot more confusing - It seems the way the C++ interface works with tensorflow is, you will have to make a query to the TF::Session object, which returns the results in the stack only, i.e. we can't do
`Tensor *t = new Tensor();`
`QueryTheTFSession(some_input, t);`
If I want to store them as vector<Tensor*>, I would have to do something like
`vector<Tensor> query_results;`
`QueryTFSession(..., &query_results); // the way TF works, you have to pass in a pointer for a vector<Tensor> and it fills this vector`
`Tensor* tensor_to_push_to_map = new Tensor(query_results[0]);`
`state_to_context.push_back(tensor_to_push_to_map)`
which I think is 1. very confusing with the manual copy from stack to heap; 2. not really necessary since from [here](https://www.tensorflow.org/versions/master/api_docs/cc/class/tensorflow/tensor) it seems a default constructor of Tensor returns an object that's almost as small as a pointer.

Or maybe I could something like the following?
`vector<Tensor>* query_results = new vector<Tensor>;`
`QueryTFSession(..., query_results);`
`state_to_context.push_back(query_results.begin()+1) // get a pointer that points to part of the allocated heap memory? I have never done this before and it feels like bad practice`

danpovey commented on src/tensorflow/Makefile:
```diff
@@ -0,0 +1,24 @@
+          $(CURDIR)/../../tools/tensorflow/bazel-bin/tensorflow/tensorflow_cc.so
+
+LDLIBS +=  -lz -ldl -fPIC -lrt
+LDLIBS += $(OTHERLIBS) -L$(TENSORFLOW)/bazel-bin/tensorflow -ltensorflow_cc
```
I don't see OTHERLIBS referred to anywhere.

danpovey commented on src/tensorflow/Makefile:
```diff
@@ -0,0 +1,24 @@
+all:
+
+EXTRA_CXXFLAGS = -Wno-sign-compare -I$(TENSORFLOW)/bazel-tensorflow/external/protobuf/src -I$(TENSORFLOW)/bazel-genfiles -I$(TENSORFLOW) -I$(TENSORFLOW)/tensorflow/contrib/makefile/downloads/eigen/
+#EXTRA_CXXFLAGS = -Wno-sign-compare -fPIC -I$(TENSORFLOW)/bazel-tensorflow/external/protobuf/src -I$(TENSORFLOW)/bazel-genfiles -I$(TENSORFLOW) -I$(TENSORFLOW)/tensorflow/contrib/makefile/downloads/eigen/
```
Should this comment still be here?

danpovey commented on egs/wsj/s5/steps/tensorflow/vanilla_rnnlm.py:
```diff
@@ -0,0 +1,319 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
```
There should be a comment (or doc-string, maybe) somewhere near the top that explains what this script does and how it's supposed to be invoked.

danpovey commented on egs/wsj/s5/steps/tensorflow/lstm_fast.py:
```diff
@@ -0,0 +1,349 @@
+# this function does the following:
+# return exp(x) if x < 0
+#        x if x >= 0
+def f(x):
```
is this used?  Please go through the code and remove anything unused, and comments that are not suitable to check in.

danpovey commented on egs/ami/s5/cmd.sh:
```diff
@@ -12,6 +12,7 @@
 
 export train_cmd="queue.pl --mem 1G"
 export decode_cmd="queue.pl --mem 2G"
+export tensorflow_cmd="queue.pl -l hostname=b*"
```
let's call this tfrnnlm_cmd, I don't want people to assume that this is about acoustic models.

danpovey commented on egs/ami/s5/local/tensorflow/prep_data.sh:
```diff
@@ -0,0 +1,69 @@
+. utils/parse_options.sh
+
+if [ $# != 1 ]; then
+   echo "Usage: $0 [options] <dest-dir>"
```
There needs to be an explanation of what this script is for and what it does.  The same goes for any other scripts that you are adding.

pzelasko commented on src/tensorflow/tensorflow-rnnlm-lib.cc:
```diff
@@ -0,0 +1,361 @@
+  checkpointPathTensor.scalar<std::string>()() = tf_model_path;
+  
+  status = session_->Run(
+      {{ graph_def.saver_def().filename_tensor_name(), checkpointPathTensor },},
```
It's a [list initialization](http://en.cppreference.com/w/cpp/language/list_initialization), the type of the initialized object is inferred from the called function signature - seems valid C++11, although the use of the comma operator (the one before the last) makes this look very strange (not sure about whether that comma is needed though).

danpovey commented:
The concern is about whether, when you copy a Tensor from one array to
another while resizing, it has to copy the memory contents over.  I would
have thought it would do this.

I don't fully understand your reply but I suspect you have misunderstood
something and you are making it more complicated than it needs to be.



On Tue, Jul 25, 2017 at 11:38 AM, hainan-xv <notifications@github.com>
wrote:

> *@hainan-xv* commented on this pull request.
> ------------------------------
>
> In src/tensorflow/tensorflow-rnnlm-lib.h
> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958>:
>
> > +  // We cannot use "const" because the pure virtual function in the interface is
> +  // not const.
> +  virtual Weight Final(StateId s);
> +
> +  virtual bool GetArc(StateId s, Label ilabel, fst::StdArc* oarc);
> +
> + private:
> +  typedef unordered_map<std::vector<Label>,
> +                        StateId, VectorHasher<Label> > MapType;
> +  StateId start_state_;
> +  MapType wseq_to_state_;
> +  std::vector<std::vector<Label> > state_to_wseq_;
> +
> +  KaldiTfRnnlmWrapper *rnnlm_;
> +  int32 max_ngram_order_;
> +  std::vector<tensorflow::Tensor> state_to_context_;
>
> I think I know why you're suggesting this, but it seems in the case of TF
> it'll make the code a lot more confusing - It seems the way the C++
> interface works with tensorflow is, you will have to make a query to the
> TF::Session object, which returns the results in the stack only, i.e. we
> can't do
> Tensor *t = new Tensor();
> QueryTheTFSession(some_input, t);
> If I want to store them as vector<Tensor*>, I would have to do something
> like
> vector<Tensor> query_results;
> QueryTFSession(..., &query_results); // the way TF works, you have to pass
> in a pointer for a vector<Tensor> and it fills this vector
> Tensor* tensor_to_push_to_map = new Tensor(query_results[0]);
> state_to_context.push_back(tensor_to_push_to_map)
> which I think is 1. very confusing with the manual copy from stack to
> heap; 2. not really necessary since from here
> <https://www.tensorflow.org/versions/master/api_docs/cc/class/tensorflow/tensor>
> it seems a default constructor of Tensor returns an object that's almost as
> small as a pointer.
>
> Or maybe I could something like the following?
> vector<Tensor>* query_results = new vector<Tensor>;
> QueryTFSession(..., query_results);
> state_to_context.push_back(query_results.begin()+1) // get a pointer that
> points to part of the allocated heap memory? I have never done this before
> and it feels like bad practice
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu5jPLXyx52QXRl62c8fb06oxJRLoks5sRjYzgaJpZM4OO-zM>
> .
>

danpovey commented:
The code changes you need to make are very simple.  Instead of doing:

```
state_to_context_.push_back(initial_context);
state_to_cell_.push_back(initial_cell);
```
you do:
```
 state_to_context_.push_back(new Tensor(initial_context));
 state_to_cell_.push_back(new Tensor(initial_cell));
```
and replace all instances of, e.g. `state_to_context_[state]` with
`*(state_to_context_[state])`.

On Tue, Jul 25, 2017 at 11:58 AM, Daniel Povey <dpovey@gmail.com> wrote:

> The concern is about whether, when you copy a Tensor from one array to
> another while resizing, it has to copy the memory contents over.  I would
> have thought it would do this.
>
> I don't fully understand your reply but I suspect you have misunderstood
> something and you are making it more complicated than it needs to be.
>
>
>
> On Tue, Jul 25, 2017 at 11:38 AM, hainan-xv <notifications@github.com>
> wrote:
>
>> *@hainan-xv* commented on this pull request.
>> ------------------------------
>>
>> In src/tensorflow/tensorflow-rnnlm-lib.h
>> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958>:
>>
>> > +  // We cannot use "const" because the pure virtual function in the interface is
>> +  // not const.
>> +  virtual Weight Final(StateId s);
>> +
>> +  virtual bool GetArc(StateId s, Label ilabel, fst::StdArc* oarc);
>> +
>> + private:
>> +  typedef unordered_map<std::vector<Label>,
>> +                        StateId, VectorHasher<Label> > MapType;
>> +  StateId start_state_;
>> +  MapType wseq_to_state_;
>> +  std::vector<std::vector<Label> > state_to_wseq_;
>> +
>> +  KaldiTfRnnlmWrapper *rnnlm_;
>> +  int32 max_ngram_order_;
>> +  std::vector<tensorflow::Tensor> state_to_context_;
>>
>> I think I know why you're suggesting this, but it seems in the case of TF
>> it'll make the code a lot more confusing - It seems the way the C++
>> interface works with tensorflow is, you will have to make a query to the
>> TF::Session object, which returns the results in the stack only, i.e. we
>> can't do
>> Tensor *t = new Tensor();
>> QueryTheTFSession(some_input, t);
>> If I want to store them as vector<Tensor*>, I would have to do something
>> like
>> vector<Tensor> query_results;
>> QueryTFSession(..., &query_results); // the way TF works, you have to
>> pass in a pointer for a vector<Tensor> and it fills this vector
>> Tensor* tensor_to_push_to_map = new Tensor(query_results[0]);
>> state_to_context.push_back(tensor_to_push_to_map)
>> which I think is 1. very confusing with the manual copy from stack to
>> heap; 2. not really necessary since from here
>> <https://www.tensorflow.org/versions/master/api_docs/cc/class/tensorflow/tensor>
>> it seems a default constructor of Tensor returns an object that's almost as
>> small as a pointer.
>>
>> Or maybe I could something like the following?
>> vector<Tensor>* query_results = new vector<Tensor>;
>> QueryTFSession(..., query_results);
>> state_to_context.push_back(query_results.begin()+1) // get a pointer
>> that points to part of the allocated heap memory? I have never done this
>> before and it feels like bad practice
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ADJVu5jPLXyx52QXRl62c8fb06oxJRLoks5sRjYzgaJpZM4OO-zM>
>> .
>>
>
>

hainan-xv commented:
What I mean is, the way TensorFlow interfaces work don't really let you
take advantage of this.
When doing a very on the TF session, we always need to pass in a
vector<pair<string, Tensor> >,
where the string indicates the name of the Tensor, and because of this we
will have
to do the copies anyway even if we store the map using a bunch of pointers.

E.g. the following code, even if we have initial_context_ as a pointer, we
still
need to create the "inputs" object which require a copy

197     std::vector<std::pair<string, tensorflow::Tensor> > inputs = {


198       {"Train/Model/test_word_in", bosword},


199       {"Train/Model/test_state_in", initial_context_},


200     };


201


202     status = session_->Run(inputs, {"Train/Model/test_cell_out"}, {},
&state);

On Tue, Jul 25, 2017 at 12:03 PM, Daniel Povey <notifications@github.com>
wrote:

> The code changes you need to make are very simple. Instead of doing:
>
> ```
> state_to_context_.push_back(initial_context);
> state_to_cell_.push_back(initial_cell);
> ```
> you do:
> ```
> state_to_context_.push_back(new Tensor(initial_context));
> state_to_cell_.push_back(new Tensor(initial_cell));
> ```
> and replace all instances of, e.g. `state_to_context_[state]` with
> `*(state_to_context_[state])`.
>
>
> On Tue, Jul 25, 2017 at 11:58 AM, Daniel Povey <dpovey@gmail.com> wrote:
>
> > The concern is about whether, when you copy a Tensor from one array to
> > another while resizing, it has to copy the memory contents over. I would
> > have thought it would do this.
> >
> > I don't fully understand your reply but I suspect you have misunderstood
> > something and you are making it more complicated than it needs to be.
> >
> >
> >
> > On Tue, Jul 25, 2017 at 11:38 AM, hainan-xv <notifications@github.com>
> > wrote:
> >
> >> *@hainan-xv* commented on this pull request.
> >> ------------------------------
> >>
> >> In src/tensorflow/tensorflow-rnnlm-lib.h
> >> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958>:
> >>
> >> > + // We cannot use "const" because the pure virtual function in the
> interface is
> >> + // not const.
> >> + virtual Weight Final(StateId s);
> >> +
> >> + virtual bool GetArc(StateId s, Label ilabel, fst::StdArc* oarc);
> >> +
> >> + private:
> >> + typedef unordered_map<std::vector<Label>,
> >> + StateId, VectorHasher<Label> > MapType;
> >> + StateId start_state_;
> >> + MapType wseq_to_state_;
> >> + std::vector<std::vector<Label> > state_to_wseq_;
> >> +
> >> + KaldiTfRnnlmWrapper *rnnlm_;
> >> + int32 max_ngram_order_;
> >> + std::vector<tensorflow::Tensor> state_to_context_;
> >>
> >> I think I know why you're suggesting this, but it seems in the case of
> TF
> >> it'll make the code a lot more confusing - It seems the way the C++
> >> interface works with tensorflow is, you will have to make a query to the
> >> TF::Session object, which returns the results in the stack only, i.e. we
> >> can't do
> >> Tensor *t = new Tensor();
> >> QueryTheTFSession(some_input, t);
> >> If I want to store them as vector<Tensor*>, I would have to do something
> >> like
> >> vector<Tensor> query_results;
> >> QueryTFSession(..., &query_results); // the way TF works, you have to
> >> pass in a pointer for a vector<Tensor> and it fills this vector
> >> Tensor* tensor_to_push_to_map = new Tensor(query_results[0]);
> >> state_to_context.push_back(tensor_to_push_to_map)
> >> which I think is 1. very confusing with the manual copy from stack to
> >> heap; 2. not really necessary since from here
> >> <https://www.tensorflow.org/versions/master/api_docs/cc/
> class/tensorflow/tensor>
> >> it seems a default constructor of Tensor returns an object that's
> almost as
> >> small as a pointer.
> >>
> >> Or maybe I could something like the following?
> >> vector<Tensor>* query_results = new vector<Tensor>;
> >> QueryTFSession(..., query_results);
> >> state_to_context.push_back(query_results.begin()+1) // get a pointer
> >> that points to part of the allocated heap memory? I have never done this
> >> before and it feels like bad practice
> >>
> >> —
> >> You are receiving this because you commented.
> >> Reply to this email directly, view it on GitHub
> >> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958>,
> or mute
> >> the thread
> >> <https://github.com/notifications/unsubscribe-auth/
> ADJVu5jPLXyx52QXRl62c8fb06oxJRLoks5sRjYzgaJpZM4OO-zM>
> >> .
> >>
> >
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-317839109>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFMCDmubJ7sYm-_X31LK5cxFXgTYZmzJks5sRjvngaJpZM4OO-zM>
> .
>



-- 
- Hainan

danpovey commented:
That can't be helped, but it's a separate issue and it's not related to the
reason I asked you to make those things pointers.  The point is that when
you push_back to a vector, the stl library may have to reallocate that
vector, which means it has to copy each variable in the vector, which for
non-trivial objects is slow.

On Tue, Jul 25, 2017 at 1:09 PM, hainan-xv <notifications@github.com> wrote:

> What I mean is, the way TensorFlow interfaces work don't really let you
> take advantage of this.
> When doing a very on the TF session, we always need to pass in a
> vector<pair<string, Tensor> >,
> where the string indicates the name of the Tensor, and because of this we
> will have
> to do the copies anyway even if we store the map using a bunch of pointers.
>
> E.g. the following code, even if we have initial_context_ as a pointer, we
> still
> need to create the "inputs" object which require a copy
>
> 197 std::vector<std::pair<string, tensorflow::Tensor> > inputs = {
>
>
> 198 {"Train/Model/test_word_in", bosword},
>
>
> 199 {"Train/Model/test_state_in", initial_context_},
>
>
> 200 };
>
>
> 201
>
>
> 202 status = session_->Run(inputs, {"Train/Model/test_cell_out"}, {},
> &state);
>
> On Tue, Jul 25, 2017 at 12:03 PM, Daniel Povey <notifications@github.com>
> wrote:
>
> > The code changes you need to make are very simple. Instead of doing:
> >
> > ```
> > state_to_context_.push_back(initial_context);
> > state_to_cell_.push_back(initial_cell);
> > ```
> > you do:
> > ```
> > state_to_context_.push_back(new Tensor(initial_context));
> > state_to_cell_.push_back(new Tensor(initial_cell));
> > ```
> > and replace all instances of, e.g. `state_to_context_[state]` with
> > `*(state_to_context_[state])`.
> >
> >
> > On Tue, Jul 25, 2017 at 11:58 AM, Daniel Povey <dpovey@gmail.com> wrote:
> >
> > > The concern is about whether, when you copy a Tensor from one array to
> > > another while resizing, it has to copy the memory contents over. I
> would
> > > have thought it would do this.
> > >
> > > I don't fully understand your reply but I suspect you have
> misunderstood
> > > something and you are making it more complicated than it needs to be.
> > >
> > >
> > >
> > > On Tue, Jul 25, 2017 at 11:38 AM, hainan-xv <notifications@github.com>
> > > wrote:
> > >
> > >> *@hainan-xv* commented on this pull request.
> > >> ------------------------------
> > >>
> > >> In src/tensorflow/tensorflow-rnnlm-lib.h
> > >> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958>:
>
> > >>
> > >> > + // We cannot use "const" because the pure virtual function in the
> > interface is
> > >> + // not const.
> > >> + virtual Weight Final(StateId s);
> > >> +
> > >> + virtual bool GetArc(StateId s, Label ilabel, fst::StdArc* oarc);
> > >> +
> > >> + private:
> > >> + typedef unordered_map<std::vector<Label>,
> > >> + StateId, VectorHasher<Label> > MapType;
> > >> + StateId start_state_;
> > >> + MapType wseq_to_state_;
> > >> + std::vector<std::vector<Label> > state_to_wseq_;
> > >> +
> > >> + KaldiTfRnnlmWrapper *rnnlm_;
> > >> + int32 max_ngram_order_;
> > >> + std::vector<tensorflow::Tensor> state_to_context_;
> > >>
> > >> I think I know why you're suggesting this, but it seems in the case of
> > TF
> > >> it'll make the code a lot more confusing - It seems the way the C++
> > >> interface works with tensorflow is, you will have to make a query to
> the
> > >> TF::Session object, which returns the results in the stack only, i.e.
> we
> > >> can't do
> > >> Tensor *t = new Tensor();
> > >> QueryTheTFSession(some_input, t);
> > >> If I want to store them as vector<Tensor*>, I would have to do
> something
> > >> like
> > >> vector<Tensor> query_results;
> > >> QueryTFSession(..., &query_results); // the way TF works, you have to
> > >> pass in a pointer for a vector<Tensor> and it fills this vector
> > >> Tensor* tensor_to_push_to_map = new Tensor(query_results[0]);
> > >> state_to_context.push_back(tensor_to_push_to_map)
> > >> which I think is 1. very confusing with the manual copy from stack to
> > >> heap; 2. not really necessary since from here
> > >> <https://www.tensorflow.org/versions/master/api_docs/cc/
> > class/tensorflow/tensor>
> > >> it seems a default constructor of Tensor returns an object that's
> > almost as
> > >> small as a pointer.
> > >>
> > >> Or maybe I could something like the following?
> > >> vector<Tensor>* query_results = new vector<Tensor>;
> > >> QueryTFSession(..., query_results);
> > >> state_to_context.push_back(query_results.begin()+1) // get a pointer
> > >> that points to part of the allocated heap memory? I have never done
> this
> > >> before and it feels like bad practice
> > >>
> > >> —
> > >> You are receiving this because you commented.
> > >> Reply to this email directly, view it on GitHub
> > >> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958>,
> > or mute
> > >> the thread
> > >> <https://github.com/notifications/unsubscribe-auth/
> > ADJVu5jPLXyx52QXRl62c8fb06oxJRLoks5sRjYzgaJpZM4OO-zM>
> > >> .
> > >>
> > >
> > >
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-317839109>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AFMCDmubJ7sYm-_
> X31LK5cxFXgTYZmzJks5sRjvngaJpZM4OO-zM>
> > .
> >
>
>
>
> --
> - Hainan
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-317857017>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuw59yMr5mgOc5HociETtTZIwv61nks5sRktigaJpZM4OO-zM>
> .
>

hainan-xv commented:
FYI the commit I just pushed isn't final yet. I'm still making some other
changes.

On Tue, Jul 25, 2017 at 1:24 PM, Daniel Povey <notifications@github.com>
wrote:

> That can't be helped, but it's a separate issue and it's not related to the
> reason I asked you to make those things pointers. The point is that when
> you push_back to a vector, the stl library may have to reallocate that
> vector, which means it has to copy each variable in the vector, which for
> non-trivial objects is slow.
>
>
> On Tue, Jul 25, 2017 at 1:09 PM, hainan-xv <notifications@github.com>
> wrote:
>
> > What I mean is, the way TensorFlow interfaces work don't really let you
> > take advantage of this.
> > When doing a very on the TF session, we always need to pass in a
> > vector<pair<string, Tensor> >,
> > where the string indicates the name of the Tensor, and because of this we
> > will have
> > to do the copies anyway even if we store the map using a bunch of
> pointers.
> >
> > E.g. the following code, even if we have initial_context_ as a pointer,
> we
> > still
> > need to create the "inputs" object which require a copy
> >
> > 197 std::vector<std::pair<string, tensorflow::Tensor> > inputs = {
> >
> >
> > 198 {"Train/Model/test_word_in", bosword},
> >
> >
> > 199 {"Train/Model/test_state_in", initial_context_},
> >
> >
> > 200 };
> >
> >
> > 201
> >
> >
> > 202 status = session_->Run(inputs, {"Train/Model/test_cell_out"}, {},
> > &state);
> >
> > On Tue, Jul 25, 2017 at 12:03 PM, Daniel Povey <notifications@github.com
> >
> > wrote:
> >
> > > The code changes you need to make are very simple. Instead of doing:
> > >
> > > ```
> > > state_to_context_.push_back(initial_context);
> > > state_to_cell_.push_back(initial_cell);
> > > ```
> > > you do:
> > > ```
> > > state_to_context_.push_back(new Tensor(initial_context));
> > > state_to_cell_.push_back(new Tensor(initial_cell));
> > > ```
> > > and replace all instances of, e.g. `state_to_context_[state]` with
> > > `*(state_to_context_[state])`.
> > >
> > >
> > > On Tue, Jul 25, 2017 at 11:58 AM, Daniel Povey <dpovey@gmail.com>
> wrote:
> > >
> > > > The concern is about whether, when you copy a Tensor from one array
> to
> > > > another while resizing, it has to copy the memory contents over. I
> > would
> > > > have thought it would do this.
> > > >
> > > > I don't fully understand your reply but I suspect you have
> > misunderstood
> > > > something and you are making it more complicated than it needs to be.
> > > >
> > > >
> > > >
> > > > On Tue, Jul 25, 2017 at 11:38 AM, hainan-xv <
> notifications@github.com>
> > > > wrote:
> > > >
> > > >> *@hainan-xv* commented on this pull request.
> > > >> ------------------------------
> > > >>
> > > >> In src/tensorflow/tensorflow-rnnlm-lib.h
> > > >> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958
> >:
> >
> > > >>
> > > >> > + // We cannot use "const" because the pure virtual function in
> the
> > > interface is
> > > >> + // not const.
> > > >> + virtual Weight Final(StateId s);
> > > >> +
> > > >> + virtual bool GetArc(StateId s, Label ilabel, fst::StdArc* oarc);
> > > >> +
> > > >> + private:
> > > >> + typedef unordered_map<std::vector<Label>,
> > > >> + StateId, VectorHasher<Label> > MapType;
> > > >> + StateId start_state_;
> > > >> + MapType wseq_to_state_;
> > > >> + std::vector<std::vector<Label> > state_to_wseq_;
> > > >> +
> > > >> + KaldiTfRnnlmWrapper *rnnlm_;
> > > >> + int32 max_ngram_order_;
> > > >> + std::vector<tensorflow::Tensor> state_to_context_;
> > > >>
> > > >> I think I know why you're suggesting this, but it seems in the case
> of
> > > TF
> > > >> it'll make the code a lot more confusing - It seems the way the C++
> > > >> interface works with tensorflow is, you will have to make a query to
> > the
> > > >> TF::Session object, which returns the results in the stack only,
> i.e.
> > we
> > > >> can't do
> > > >> Tensor *t = new Tensor();
> > > >> QueryTheTFSession(some_input, t);
> > > >> If I want to store them as vector<Tensor*>, I would have to do
> > something
> > > >> like
> > > >> vector<Tensor> query_results;
> > > >> QueryTFSession(..., &query_results); // the way TF works, you have
> to
> > > >> pass in a pointer for a vector<Tensor> and it fills this vector
> > > >> Tensor* tensor_to_push_to_map = new Tensor(query_results[0]);
> > > >> state_to_context.push_back(tensor_to_push_to_map)
> > > >> which I think is 1. very confusing with the manual copy from stack
> to
> > > >> heap; 2. not really necessary since from here
> > > >> <https://www.tensorflow.org/versions/master/api_docs/cc/
> > > class/tensorflow/tensor>
> > > >> it seems a default constructor of Tensor returns an object that's
> > > almost as
> > > >> small as a pointer.
> > > >>
> > > >> Or maybe I could something like the following?
> > > >> vector<Tensor>* query_results = new vector<Tensor>;
> > > >> QueryTFSession(..., query_results);
> > > >> state_to_context.push_back(query_results.begin()+1) // get a
> pointer
> > > >> that points to part of the allocated heap memory? I have never done
> > this
> > > >> before and it feels like bad practice
> > > >>
> > > >> —
> > > >> You are receiving this because you commented.
> > > >> Reply to this email directly, view it on GitHub
> > > >> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r129389958
> >,
> > > or mute
> > > >> the thread
> > > >> <https://github.com/notifications/unsubscribe-auth/
> > > ADJVu5jPLXyx52QXRl62c8fb06oxJRLoks5sRjYzgaJpZM4OO-zM>
> > > >> .
> > > >>
> > > >
> > > >
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-317839109>,
> > or mute
> > > the thread
> > > <https://github.com/notifications/unsubscribe-auth/AFMCDmubJ7sYm-_
> > X31LK5cxFXgTYZmzJks5sRjvngaJpZM4OO-zM>
> > > .
> > >
> >
> >
> >
> > --
> > - Hainan
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-317857017>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ADJVuw59yMr5mgOc5HociETtTZIwv61nks5sRktigaJpZM4OO-zM>
>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-317861188>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFMCDs1qTgWrhj8nT9WrSPNXkfBYLAQ9ks5sRk8FgaJpZM4OO-zM>
> .
>



-- 
- Hainan

vince62s commented:
Hainan,
while you're still working on this. Just be aware that the sampled softmax code is very inefficient (from the TF tuto) because they never changed some code which uses the transpose function.
see this https://github.com/tensorflow/tensorflow/issues/5350 and this https://github.com/tensorflow/tensorflow/issues/4138
When ever you use a big vocab (>100k) it makes things very slow (almost as slow as full softmax). I can confirm when we tested your PR.
We used a trick to save the transpose weights instead and then when training is finished we used a small code to convert the transposed weights, to be used for inference.

danpovey commented:
I don't think there is an immediate plan to support sampling for the
TF-based setup.  The nnet3-based setup will support that.  For now, I
think, the TF-based setup is using a small vocabulary.

On Wed, Jul 26, 2017 at 11:25 AM, vince62s <notifications@github.com> wrote:

> Hainan,
> while you're still working on this. Just be aware that the sampled softmax
> code is very inefficient (from the TF tuto) because they never changed some
> code which uses the transpose function.
> see this tensorflow/tensorflow#5350
> <https://github.com/tensorflow/tensorflow/issues/5350> and this
> tensorflow/tensorflow#4138
> <https://github.com/tensorflow/tensorflow/issues/4138>
> When ever you use a big vocab (>100k) it makes things very slow (almost as
> slow as full softmax). I can confirm when we tested your PR.
> We used a trick to save the transpose weights instead and then when
> training is finished we used a small code to convert the transposed
> weights, to be used for inference.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-318140807>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuzw8i1-sQMAM8OUwWMKRcuTWSxsrks5sR4ShgaJpZM4OO-zM>
> .
>

hainan-xv commented:
I fixed all the issues mentioned and re-tested the code on real data, as well as making sure valgrind does not complain.

danpovey commented:
Thanks.  I'll try to re-review to-day or to-morrow.

On Wed, Jul 26, 2017 at 4:30 PM, hainan-xv <notifications@github.com> wrote:

> I fixed all the issues mentioned and re-tested the code on real data, as
> well as making sure valgrind does not complain.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-318212515>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuxCB6HFDYa3JJLur0vtioT_Lpzngks5sR8v6gaJpZM4OO-zM>
> .
>

danpovey commented on egs/ami/s5/local/tfrnnlm/rnnlm_data_prep.sh:
```diff
@@ -0,0 +1,67 @@
+
+total_nwords=`wc -l $dir/unigram.counts | awk '{print $1}'`
+
+head -$nwords $dir/unigram.counts | awk '{print $2}' | tee $dir/wordlist.rnn | awk '{print NR-1, $1}' > $dir/wordlist.rnn.id
```
What is this wordlist.rnn.id about?  It needs some explanation.

danpovey commented on egs/ami/s5/local/tfrnnlm/rnnlm_data_prep.sh:
```diff
@@ -0,0 +1,67 @@
+head -$nwords $dir/unigram.counts | awk '{print $2}' | tee $dir/wordlist.rnn | awk '{print NR-1, $1}' > $dir/wordlist.rnn.id
+tail -n +$nwords $dir/unigram.counts > $dir/unk_class.counts
+
+for type in train valid; do
```
need a comment here explaining what's happening.

danpovey commented on egs/ami/s5/local/tfrnnlm/rnnlm_data_prep.sh:
```diff
@@ -0,0 +1,67 @@
+cat $dir/unk_class.counts | awk '{print $2, $1}' > $dir/unk.probs
+cp $dir/wordlist.rnn $dir/wordlist.rnn.final
+
+has_oos=`grep "<oos>" $dir/wordlist.rnn.final | wc -l | awk '{print $1}'`
```
you can do
```
if grep -w '<oos>' $dir/wordlist.rnn.final >/dev/null; then
  ...
```

danpovey commented on egs/ami/s5/local/tfrnnlm/run_lstm.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+mic=ihm
+ngram_order=4
```
explain that this affects state-merging when rescoring (an approximation for speed).

danpovey commented on egs/ami/s5/local/tfrnnlm/run_lstm.sh:
```diff
@@ -0,0 +1,45 @@
+mic=ihm
+ngram_order=4
+stage=1
+weight=0.5
```
explain what the weight is about.

danpovey commented on egs/ami/s5/local/tfrnnlm/run_lstm.sh:
```diff
@@ -0,0 +1,45 @@
+
+mkdir -p $dir
+if [ $stage -le 2 ]; then
+# the following script uses TensorFlow. You could use tools/extras/install_tensorflow_py.sh to install it
```
This script doesn't check if TensorFlow is installed. If it's not installed it should die with a message explaining how to install it.   This script should also check whether the TensorFlow-related binaries have been compiled, and if not, explain how to compile them.

danpovey commented on egs/ami/s5/local/tfrnnlm/run_lstm.sh:
```diff
@@ -0,0 +1,45 @@
+mkdir -p $dir
+if [ $stage -le 2 ]; then
+# the following script uses TensorFlow. You could use tools/extras/install_tensorflow_py.sh to install it
+  python steps/tfrnnlm/rnnlm.py --data_path=$dir --save_path=$dir/rnnlm --vocab_path=$dir/wordlist.rnn.final
```
Since this script has lstm in its name, shouldn't it be explicitly calling the  lstm version of the script?
Please try to use "-" instead of "_" in options to python scripts, if possible, for consistency with the rest of our Python scripts.

danpovey replied:
I don't understand how this example script is supposed to interact with the queue.  Shouldn't it be reserving a GPU to run rnnlm.py?    It would't make sense for this entire script to be invoked on the queue with a GPU, because then you'd be wasting the GPU when running the other stages, including lattice rescoring; and any error messages would be lost.

danpovey commented on egs/wsj/s5/steps/tfrnnlm/lstm.py:
```diff
@@ -0,0 +1,328 @@
+# limitations under the License.
+# ==============================================================================
+
+# this script trains a vanilla RNNLM with TensorFlow. 
```
I assume this script takes all the GPUs that it can see, if compiled for GPU?
On our queue, this is a problem; you have to use Gaurav's script  to get a suitable number of devices, and set CUDA_VISIBLE_DEVICES appropriately.  I've heard that TensorFlow does not actually make use of additional GPUs (beyond one) very efficiently, so please experiment and see whether we actually want to use more than one GPU.

Also you should explain in this script that it will use all visible GPUs and you have to set CUDA_VISIBLE_DEVICES (e.g. to "1", or "0,2", i.e. the subset of devices you want it to restrict its use to), if you don't want it to use all visible GPUs.  This is obvious to people who know TensorFlow but may not be obvious to Kaldi users.

danpovey commented on egs/wsj/s5/steps/tfrnnlm/vanilla_rnnlm.py:
```diff
@@ -0,0 +1,326 @@
+    with tf.name_scope("Train"):
+      train_input = RNNLMInput(config=config, data=train_data, name="TrainInput")
+      with tf.variable_scope("Model", reuse=None, initializer=initializer):
+        m = RNNLMModel(is_training=True, config=config, input_=train_input)
```
in class names, please use RnnlmModel etc., don't use all-caps.

danpovey commented on src/tfrnnlm/Makefile:
```diff
@@ -0,0 +1,34 @@
+# Tis directory is not compiled automatically when you do make under src/
```
typo

danpovey commented on src/tfrnnlm/tensorflow-rnnlm.cc:
```diff
@@ -0,0 +1,376 @@
+// tensorflow-rnnlm-lib.cc
```
name needs to be fixed here.

danpovey commented on src/tfrnnlm/tensorflow-rnnlm.cc:
```diff
@@ -0,0 +1,376 @@
+  for (int32 i = 0; i < fst_label_to_word_.size(); ++i) {
+    fst_label_to_word_[i] = fst_word_symbols->Find(i);
+    if (fst_label_to_word_[i] == "") {
+      KALDI_ERR << "Could not find word for integer " << i << "in the word "
```
space before 'in'

danpovey commented on src/tfrnnlm/tensorflow-rnnlm.cc:
```diff
@@ -0,0 +1,376 @@
+
+      int fst_label = fst_word_symbols->Find(word);
+      if (fst::SymbolTable::kNoSymbol == fst_label) {
+        if (id == eos_) {
```
unnecessary braces and whitespace.

danpovey commented on src/tfrnnlm/tensorflow-rnnlm.h:
```diff
@@ -0,0 +1,179 @@
+//
+class KaldiTfRnnlmWrapper {
+ public:
+  // usually FSTs and the RNNLM would have different wordlists so we have 2
```
Try to document this better, and use Doxygen-style documentation for most functions , documenting
their parameters.  This comment that you have here is more like a sequence of asides clarifications, it doesn't really clearly explain
from the top, what's going on, and doesn't make it easy to look up for a particular parameter, what it
is.  For inspiration, look at my PR about RNNLM stuff, and look for '@' which appears in Doxygen-style
comments.  Note: `///` and `/**` are used to start Doxygen comment blocks.

danpovey commented on src/tfrnnlm/tensorflow-rnnlm.h:
```diff
@@ -0,0 +1,179 @@
+
+  void Register(OptionsItf *opts) {
+    opts->Register("unk-symbol", &unk_symbol, "Symbol for out-of-vocabulary "
+                   "words in rnnlm. (default = <oos>)");
```
Defaults will be printed anyway so you don't need to write them here.

danpovey commented on src/tfrnnlm/tensorflow-rnnlm.h:
```diff
@@ -0,0 +1,179 @@
+  }
+};
+
+// this class wrapps the tensorflow based RNNLM and provides a set of interfaces
```
typo.
Also all comments should be properly capitalized and punctuated.

danpovey commented on src/tfrnnlm/tensorflow-rnnlm.cc:
```diff
@@ -0,0 +1,376 @@
+#include "util/text-utils.h"
+
+namespace kaldi {
+using std::ifstream;
```
Run cpplint.py and see if it complains about these using statements/declarations/directives/whatever, I'm not sure if they're allowed at file scope.

hainan-xv replied:
Just tried cpplint and it seems
using some_scope::some_function is OK.

cpplint is actually complaining about thing like
"using namespace kaldi", citing "Do not use namespace using-directives.  Use using-declarations instead.", but I guess we're already using that in probably every kaldi binary, so maybe just keep it as it is.

danpovey commented on src/tfrnnlmbin/Makefile:
```diff
@@ -0,0 +1,36 @@
+# Tis directory is not compiled automatically when you do make under src/
```
typo tis.

danpovey commented on src/tfrnnlmbin/Makefile:
```diff
@@ -0,0 +1,36 @@
+          $(shell pwd)/../../tools/tensorflow/bazel-bin/tensorflow/libtensorflow_cc.so
+
+LDLIBS +=  -lz -ldl -fPIC -lrt
+#LDLIBS += -rpath,$(shell pwd)/../../tools/tensorflow/bazel-bin/tensorflow/
```
I'm surprised that the rpath thing was not required.  Make sure that this binary can be run from a normal command line without sourcing the TensorFlow virtual environment.

hainan-xv replied:
Oh I actually have rpath in EXTRA_CXXFLAGS
Do you prefer if it is here in LDLIBS?

danpovey replied:
Try adding it to LDFLAGS, e.g. using
`LDFLAGS += ....`
that's where it belongs.
Break up EXTRA_CXXFLAGS with backslash if needed, I couldn't see that because it was off the screen.
And remove that comment about `LDLIBS += ...`

danpovey commented on src/tfrnnlmbin/lattice-lmrescore-tf-rnnlm.cc:
```diff
@@ -0,0 +1,142 @@
+
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 6 && po.NumArgs() != 5) {
```
swap 5 and 6.

hainan-xv commented on egs/ami/s5/local/tfrnnlm/rnnlm_data_prep.sh:
```diff
@@ -0,0 +1,67 @@
+
+cat $srcdir/lexicon.txt | awk '{print $1}' | sort -u | grep -v -w '!SIL' > $dir/wordlist.all
+
+# Get training data with OOV words (w.r.t. our current vocab) replaced with <unk>.
```
I don't understand what you mean here.

danpovey replied:
I just updated my original comment with backticks so that github doesn't treat the EOS marker as formatting and remove it.

danpovey commented:
@wantee, do you have time to try to run at least the parts of this PR that involve installing TensorFlow and compiling the extra source directory?  (Note: it's not installed by default).
I want to see how complicated the installation is for someone who is not on our grid.
I don't expect it to be as easy as installing the normally required parts of Kaldi, because installing TensorFlow is normally quite challenging, but I do want to get a sense of how much work is required and how helpful these scripts are.

hainan-xv commented:
BTW I just pushed my latest changes, having addressed the issues mentioned
before.

On Tue, Aug 1, 2017 at 3:05 PM, Daniel Povey <notifications@github.com>
wrote:

> @wantee <https://github.com/wantee>, do you have time to try to run at
> least the parts of this PR that involve installing TensorFlow and compiling
> the extra source directory? (Note: it's not installed by default).
> I want to see how complicated the installation is for someone who is not
> on our grid.
> I don't expect it to be as easy as installing the normally required parts
> of Kaldi, because installing TensorFlow is normally quite challenging, but
> I do want to get a sense of how much work is required and how helpful these
> scripts are.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#issuecomment-319510798>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFMCDtm5bjbTr35XDN_ZMq7nrxtBOJA8ks5sT6E_gaJpZM4OO-zM>
> .
>



-- 
- Hainan

danpovey commented on egs/ami/s5/cmd.sh:
```diff
@@ -12,6 +12,21 @@
+# we limit to certain machines because TF binaries compiled from C++ take advantage
+# of machine architectures for optimization
+
+export tfrnnlm_py_cmd="queue.pl --gpu 1 CUDA_VISIBLE_DEVICES=\`/home/gkumar/scripts/free-gpu\`"
```
Does this actually work?  It's far from clear to me at what stage the backtick command would be evaluated.  I wrote the script queue.pl, but even I don't understand how that works.   It also might not be portable to other similar wrapping scripts (alternatives to queue.pl).  A more generic solution would be to have the TF scripts called from shell scripts which would do this directly (and we should check into the Kaldi repo a version of this free-gpu script, with Gaurav's permission and with appropriate style and copyright headers, etc).  
Actually what might be most convenient from a scripting perspective is to write a script like steps/limit_num_gpus.py which when invoked as;
```
 steps/limit_num_gpus.py  foo --bar=baz
```
would set CUDA_VISIBLE_DEVICES to use just one GPU and would call
```
 foo --bar=baz
```
and when called as
```
 steps/limit_num_gpus.py  --num-gpus=2 foo --bar=baz
```
would do the same but allow 2 GPUs.  (The current script free-gpu does allow you to specify the number of GPUs).
If you rewrite substantially you won't need Gaurav's copyright permission as it will be a rewrite.

hainan-xv replied:
OK I have just completely re-written a wrapper that does the job. I will push the commits soon.

danpovey commented on egs/ami/s5/local/tfrnnlm/RESULTS:
```diff
@@ -0,0 +1,25 @@
+%WER 24.5 | 12643 89977 | 78.4 13.4 8.2 2.9 24.5 55.6 | -0.164 | exp/ihm/nnet3/tdnn_sp/decode_eval.rnnlm.ce.cued.lat.4gram/ascore_11/eval_hires.ctm.filt.sys
+
+Mikolov-RNNLM:
+%WER 23.9 | 13098 94471 | 79.4 12.0 8.6 3.3 23.9 57.1 | -0.188 | exp/ihm/nnet3/tdnn_sp/decode_dev.rnnlm.mik.lat.4gram/ascore_12/dev.ctm.filt.sys
```
i suspect you haven't done much work to tune this.  normally it would be the best.

hainan-xv replied:
Hmm the Mikolov script was not written by me (I think Guoguo wrote it) and I have done minimal tuning of CUED-RNNLM when comparing with Mikolov's. Also from the CUED-RNNLM paper, they reported of having around 0.5 - 0.7 gain over Mikolov's RNNLM when performing 50-best rescoring (Table 1) http://mi.eng.cam.ac.uk/~xl207/publications/conferences/ICASSP16-cued-rnnlm.pdf

danpovey commented on egs/ami/s5/local/tfrnnlm/RESULTS:
```diff
@@ -0,0 +1,25 @@
+%WER 23.4 | 13098 94466 | 79.7 11.6 8.7 3.1 23.4 56.8 | -0.213 | exp/ihm/nnet3/tdnn_sp/decode_dev.unk.fast.tfrnnlm.lat.3gram.0.5/ascore_11/dev_hires.ctm.filt.sys
+%WER 24.6 | 12643 89965 | 78.2 13.3 8.6 2.8 24.6 55.6 | -0.148 | exp/ihm/nnet3/tdnn_sp/decode_eval.unk.fast.tfrnnlm.lat.3gram.0.5/ascore_11/eval_hires.ctm.filt.sys
+
+%WER 23.1 | 13098 94461 | 79.9 11.5 8.6 3.0 23.1 56.5 | -0.212 | exp/ihm/nnet3/tdnn_sp/decode_dev.unk.fast.tfrnnlm.lat.4gram.0.5/ascore_11/dev_hires.ctm.filt.sys
```
what is the difference between these two sets of results?

hainan-xv replied:
One is 3-gram and the other is 4-gram. I included the 3-gram results because 3-gram LSTM seems to do as well as 4-gram Vanilla RNNLM.

danpovey commented on egs/wsj/s5/steps/tfrnnlm/lstm_fast.py:
```diff
@@ -0,0 +1,344 @@
+                    "Where the wordlist file is stored.")
+flags.DEFINE_string("save-path", None,
+                    "Model output directory.")
+flags.DEFINE_bool("use-fp16", False,
```
did you test this fp16 thing, and does it speed things up?

hainan-xv replied:
I haven't yet. It was there because the original TF script has it and I thought it might be useful.
Do you want me to delete this? Or train another system with fp16 and report its results/speed?

danpovey commented on egs/wsj/s5/steps/tfrnnlm/rnnlm.py:
```diff
@@ -0,0 +1 @@
+lstm.py
```
do you think this link is needed and helpful?   why not just call the script we want directly?

hainan-xv replied:
rm'ed.

danpovey commented on src/tfrnnlmbin/lattice-lmrescore-tf-rnnlm.cc:
```diff
@@ -0,0 +1,142 @@
+// tfrnnlmbin/lattice-lmrescore-tf-rnnlm.cc
+
```
are you checking that this is compiled from the scripts and giving a suitable message if not?  I don't want to have users asking about this on the list.

hainan-xv replied:
Yes I added a steps/tfrnnlm/check_tensorflow_installed.sh file which checks 1. TF is installed on the Python side 2. TF related binaries are compiled. It gives instructions of how to install them when the checks fail.

danpovey commented:
oh OK, maybe the CUED one really is better.


On Wed, Aug 9, 2017 at 4:42 PM, hainan-xv <notifications@github.com> wrote:

> *@hainan-xv* commented on this pull request.
> ------------------------------
>
> In egs/ami/s5/local/tfrnnlm/RESULTS
> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r132334189>:
>
> > @@ -0,0 +1,25 @@
> +Here are some numbers comparing rescoring performances of different RNNLM toolkits
> +
> +Un-rescored baseline:
> +%WER 24.2 | 13098 94476 | 79.2 12.0 8.8 3.4 24.2 57.4 | -0.157 | exp/ihm/nnet3/tdnn_sp/decode_dev/ascore_12/dev_hires.ctm.filt.sys
> +%WER 25.4 | 12643 89973 | 77.7 14.0 8.3 3.1 25.4 56.5 | -0.144 | exp/ihm/nnet3/tdnn_sp/decode_eval/ascore_11/eval_hires.ctm.filt.sys
> +
> +CUED-RNNLM with CE objf:
> +%WER 23.4 | 13098 94468 | 79.6 11.6 8.8 3.1 23.4 56.6 | -0.189 | exp/ihm/nnet3/tdnn_sp/decode_dev.rnnlm.ce.cued.lat.4gram/ascore_12/dev_hires.ctm.filt.sys
> +%WER 24.5 | 12643 89977 | 78.4 13.4 8.2 2.9 24.5 55.6 | -0.164 | exp/ihm/nnet3/tdnn_sp/decode_eval.rnnlm.ce.cued.lat.4gram/ascore_11/eval_hires.ctm.filt.sys
> +
> +Mikolov-RNNLM:
> +%WER 23.9 | 13098 94471 | 79.4 12.0 8.6 3.3 23.9 57.1 | -0.188 | exp/ihm/nnet3/tdnn_sp/decode_dev.rnnlm.mik.lat.4gram/ascore_12/dev.ctm.filt.sys
>
> Hmm the Mikolov script was not written by me (I think Guoguo wrote it) and
> I have done minimal tuning of CUED-RNNLM when comparing with Mikolov's.
> Also from the CUED-RNNLM paper, they reported of having around 0.5 - 0.7
> gain over Mikolov's RNNLM when performing 50-best rescoring (Table 1)
> http://mi.eng.cam.ac.uk/~xl207/publications/conferences/ICASSP16-cued-
> rnnlm.pdf
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r132334189>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuzX5LSyAOY0Af9tMHkdxNJIoEra6ks5sWkPlgaJpZM4OO-zM>
> .
>

danpovey commented:
It's Ok, you can keep it, I just wondered if you had tried it.
I think it only applies to newer GPUs anyway.

On Wed, Aug 9, 2017 at 4:46 PM, hainan-xv <notifications@github.com> wrote:

> *@hainan-xv* commented on this pull request.
> ------------------------------
>
> In egs/wsj/s5/steps/tfrnnlm/lstm_fast.py
> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r132334747>:
>
> > +import tensorflow as tf
> +
> +import reader
> +
> +flags = tf.flags
> +logging = tf.logging
> +
> +flags.DEFINE_integer("hidden-size", 200, "hidden dim of RNN")
> +
> +flags.DEFINE_string("data-path", None,
> +                    "Where the training/test data is stored.")
> +flags.DEFINE_string("vocab-path", None,
> +                    "Where the wordlist file is stored.")
> +flags.DEFINE_string("save-path", None,
> +                    "Model output directory.")
> +flags.DEFINE_bool("use-fp16", False,
>
> I haven't yet. It was there because the original TF script has it and I
> thought it might be useful.
> Do you want me to delete this? Or train another system with fp16 and
> report its results/speed?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r132334747>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu9zuXzBlO0GYccBNjfOiy_vq31MTks5sWkTlgaJpZM4OO-zM>
> .
>

danpovey commented on egs/wsj/s5/steps/limit_num_gpus.sh:
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+# This script functions as a wrapper of a bash command that uses GPUs.
```
this should be in utils/.  
Actually, create a new subdirectory utils/parallel/, add this script to there, move pbs.pl, run.pl, slurm.pl and queue.pl there and create soft links from utils for those existing scripts for back-compatibility.


(the norm for scripts that don't take the

hainan-xv replied:
Did something get cut off in this comment?

danpovey replied:
updated, ignore that part.

danpovey commented on egs/wsj/s5/steps/limit_num_gpus.sh:
```diff
@@ -0,0 +1,33 @@
+
+for i in `seq 0 $[$num_total_gpus-1]`; do
+# going over all GPUs and check if it is idle, and add to the list if yes
+  idle=`nvidia-smi -i $i | grep "No running processes found" | wc -l | awk '{print $1}'`
```
I'm not sure that this would be sufficiently fast if nvidia-persistenced were not running, but anyway, we'll address this later if it's an issue.  The awk here is not needed since wc is running from the stdin.
Actually you can do just do:
```
  if nvidia-smi -i $i | grep "No running processes found" >/dev/null; then
    # do something
  fi
```
instead of setting the idle variable.  grep returns 1 if it mached no lines, 0 otherwise.

danpovey commented on egs/wsj/s5/steps/limit_num_gpus.sh:
```diff
@@ -0,0 +1,33 @@
+CUDA_VISIBLE_DEVICES=`echo $CUDA_VISIBLE_DEVICES | sed "s=,$==g"`
+echo Running the job on GPU\(s\) $CUDA_VISIBLE_DEVICES
+
+CUDA_VISIBLE_DEVICES=$CUDA_VISIBLE_DEVICES $@
```
I think it would be clearer to do
export CUDA_VISIBLE_DEVICES
and then invoke
`"$@"`
The quoting is necessary if any command line args had spaces or needed quoting.

danpovey commented on egs/wsj/s5/steps/limit_num_gpus.sh:
```diff
@@ -0,0 +1,33 @@
+
+[ $num_gpus_assigned -ne $num_gpus ] && echo Could not find enough idle GPUs && exit -1
+
+CUDA_VISIBLE_DEVICES=`echo $CUDA_VISIBLE_DEVICES | sed "s=,$==g"`
```
I prefer `$()` to backticks.

danpovey commented on egs/wsj/s5/steps/limit_num_gpus.sh:
```diff
@@ -0,0 +1,33 @@
+[ $num_gpus_assigned -ne $num_gpus ] && echo Could not find enough idle GPUs && exit -1
+
+CUDA_VISIBLE_DEVICES=`echo $CUDA_VISIBLE_DEVICES | sed "s=,$==g"`
+echo Running the job on GPU\(s\) $CUDA_VISIBLE_DEVICES
```
Make this:
```
echo "$0: Running the job on GPU(s) $CUDA_VISIBLE_DEVICES"
```

danpovey commented on egs/wsj/s5/steps/limit_num_gpus.sh:
```diff
@@ -0,0 +1,33 @@
+num_gpus=1
+
+[ -f ./path.sh ] && . ./path.sh; # source the path.
+. parse_options.sh || exit 1;
```
don't use parse_options.sh for this and don't source path.sh-- too many dependencies.  Manually parse the --num-gpus option.
Check that the number of remaining command-line args after partsing --num-gpus is nonzero, and die with a usage message if not.  make sure the usage message is informative and contains an example.

danpovey commented on egs/wsj/s5/steps/limit_num_gpus.sh:
```diff
@@ -0,0 +1,33 @@
+# would automatically grabs all resources available on the system, e.g. a
+# TensorFlow program.
+
+num_gpus=1
```
add a comment to explain the meaning of this variable.

hainan-xv commented:
All issues addressed.

On Fri, Aug 11, 2017 at 1:36 PM, Daniel Povey <notifications@github.com>
wrote:

> *@danpovey* commented on this pull request.
> ------------------------------
>
> In egs/wsj/s5/steps/limit_num_gpus.sh
> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r132778487>:
>
> > @@ -0,0 +1,33 @@
> +#!/bin/bash
> +
> +# This script functions as a wrapper of a bash command that uses GPUs.
>
> updated, ignore that part.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1741#discussion_r132778487>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFMCDoLZnbrgMPyiB2lkBz7Vx39xQRMTks5sXLs8gaJpZM4OO-zM>
> .
>



-- 
- Hainan

danpovey commented on egs/wsj/s5/utils/parallel/limit_num_gpus.sh:
```diff
@@ -0,0 +1,51 @@
+           # CUDA_VISIBLE_DEVICES variable
+set -e
+
+if [ "$1" == "--num-gpus" ] || [ "$1" == "--num_gpus" ]; then
```
don't accept the underscore version.  And exit with status 1 on error, not -1.

danpovey commented on egs/wsj/s5/utils/parallel/limit_num_gpus.sh:
```diff
@@ -0,0 +1,51 @@
+fi
+
+if [ $# -eq 0 ]; then
+  echo $0: Missing commands
```
usage message should go here, e.g.
```
echo "Usage:  $0 [--num-gpus <num-gpus>] <command> [<arg1>...]"
echo "Runs <command> with args after setting CUDA_VISIBLE_DEVICES to "
echo "make sure exactly <num-gpus> GPUs are visible (default: 1)."
```
